ext_mongodb.php

```diff
@@ -19,6 +19,108 @@ function bsonUnserialize(array $data) : void;
 {
 }
 
+/* == MONITORING ============================================================= */
+namespace MongoDB\Driver\Monitoring;
+
+/* ------ Building blocks for value obejcts */
+abstract class CommandEvent
+{
+	private function __construct()
+	{
+		throw new Exception\RunTimeException("Accessing private constructor");
+	}
+
+	public function getCommandName() : string;
+
+	// do we need this? I don't think so, as you pass in an object
+	public function getContext() : mixed;
+
+	// libmongoc use getHost, but spec says "connectionId".
+	public function getHost() : \MongoDB\Driver\Server;
+
+	// Not adding getHint, as we don't really expose that.
+
+	// technically int64_t. Spec has "Id", and not "ID".
+	public function getOperationId() : integer;
+
+	// technically int64_t
+	public function getRequestId() : integer;
+}
+
+/* This is a separate one, so that one class can implement all three classes without
+ * a duplicate defined interface method */
+abstract class CommandResultEvent extends CommandEvent
+{
+	// spec is unitless, libmongoc does microseconds, Spec suggests "Nanos" as
+	// another alternative. (technically int64_t)
+	public function getDurationMS() : integer;
+}
+
+/* ------ Value Objects ------------------- */
+final class CommandStartedEvent extends CommandEvent
+{
+	// raw command document, with default BSON conversion rules
+	public function getCommand() : object;
+
+	public function getDatabaseName() : string;
+}
+
+final class CommandSucceededEvent extends CommandResultEvent
+{
+	// raw command reply, with default conversion rules
+	public function getReply() : object;
+}
+
+final class CommandFailedEvent extends CommandResultEvent
+{
+	// Spec calls it failure, libmongoc returns a bson_error_t.
+	public function getError() : \MongoDB\Driver\Exception\Exception;
+}
+
+/* ------ Subscriber Interface ------------ */
+interface Subscriber
+{
+	public function commandStarted( \MongoDB\Driver\Monitoring\CommandStartedEvent $event );
+	public function commandSucceeded( \MongoDB\Driver\Monitoring\CommandSucceededEvent $event );
+	public function commandFailed( \MongoDB\Driver\Monitoring\CommandFailedEvent $event );
+}
+
+/* Methods to add on Manager: */
+/*
+	function subscribe( \MongoDB\Driver\Monitoring\Subscriber $subscriber )
```

Comment:
I would like to see `addSubscriber()` and `removeSubscriber()` methods. Internally, this should be a matter of collecting the subscribers in an array and looping over the current subscribers at the time an event is dispatched.

For prior art, I'll suggest [Doctrine\Common\EventManager](https://github.com/doctrine/common/blob/master/lib/Doctrine/Common/EventManager.php). The specifics about dispatching and event listeners (i.e. individual callbacks for named event strings) are not relevant to our use case, but do consider the `addEventSubscriber()` and `removeEventSubscriber()`.

In Doctrine's case, subscribers also have a method that returns the events for which they are interested in listening, and the manager simply registers the subscriber's callback for each named event. That isn't relevant to us, since our subscriber interface is an all-in approach. If users don't care about a `commandFailed()` event, they can simply define it and do nothing within the implementation.

[PSR-14: Event Manager](https://github.com/php-fig/fig-standards/blob/master/proposed/event-manager.md) also bears mentioning, although it's far from complete and does not deal with subscribers. It does, however, include the use case of adding and removing listeners -- just to get the point across that adding/removing is a common use case.